File: 2023/erlang/day04a/day04a.erl
-module (day04a).
-export ([main/1]).

usage() ->
    Module = "day04a.erl",
    io:format("usage: escript ~p <file>~n", [Module]),
    halt(1).

process_input(Contents) ->
    lists:foldl(fun (Line, Result) ->
        Rest = hd(tl(string:split(Line, ": "))),
        WinningStr = hd(string:split(Rest, " | ")),
        WinningSet = lists:foldl(fun (NumStr, PrevSet) ->
            if NumStr /= "" ->
                sets:add_element(string:to_integer(NumStr), PrevSet)
            ; true ->
                PrevSet
            end
        end, sets:new(), string:split(WinningStr, " ", all)),
        HandStr = hd(tl(string:split(Rest, " | "))),
        HandSet = lists:foldl(fun (NumStr, PrevSet) ->
            if NumStr /= "" ->
                sets:add_element(string:to_integer(NumStr), PrevSet)
            ; true ->
                PrevSet
            end
        end, sets:new(), string:split(HandStr, " ", all)),
        Intersection = sets:intersection(WinningSet, HandSet),
        Count = sets:size(Intersection),
        if Count /= 0 ->
            Result + (1 bsl (Count - 1))
        ; true ->
            Result
        end
    end, 0, string:split(string:trim(Contents), "\n", all)).

main([Filename]) ->
    {ok, File} = file:read_file(Filename),
    Contents = unicode:characters_to_list(File),
    Result = process_input(Contents),
    io:format("result = ~w~n", [Result]);
main(_) ->
    usage().

File: 2023/elixir/day04a/day04a.exs
import Bitwise

defmodule Day04a do
  @progname "day04a.exs"

  def usage do
    IO.puts("usage: #{@progname} <file>")
    System.halt(1)
  end

  def process(contents) do
    lines = String.split(contents, "\n", trim: true)
    Enum.reduce(lines, 0, fn line, result ->
        rest = hd(tl(String.split(line, ~r{:\s+})))
        winning_str = hd(String.split(rest, ~r{\s+\|\s+}))
        winning_set = Enum.reduce(winning_str |> String.split(~r{\s+}) |> Enum.map(&String.to_integer/1),
                                  MapSet.new(),
                                  fn num, prev_set ->
                                      MapSet.put(prev_set, num)
                                  end)
        hand_str = hd(tl(String.split(rest, ~r{\s+\|\s+})))
        hand_set = Enum.reduce(hand_str |> String.split(~r{\s+}) |> Enum.map(&String.to_integer/1),
                                  MapSet.new(),
                                  fn num, prev_set ->
                                      MapSet.put(prev_set, num)
                                  end)
        count = MapSet.intersection(winning_set, hand_set) |> MapSet.size
        if count > 0 do
          result + (1 <<< (count - 1))
        else
          result
        end
    end)
  end

  def main do
    if length(System.argv()) < 1 do
      usage()
    end

    filename = hd(System.argv())
    {_, contents} = File.read(filename)
    result = process(contents)
    IO.puts("result = #{result}")
  end

end

Day04a.main()
